docs/dev/lsp-extensions.md

```diff
@@ -311,6 +311,50 @@ Moreover, it would be cool if editors didn't need to implement even basic langua
   This is how `SelectionRange` request works.
 * Alternatively, should we perhaps flag certain `SelectionRange`s as being brace pairs?
 
+## Runnables
+
+**Issue:** https://github.com/microsoft/language-server-protocol/issues/944
+
+**Server Capability:** `{ "runnables": { "kinds": string[] } }`
+
+This request is send from client to server to get the list of things that can be run (tests, binaries, `cargo check -p`).
+
+**Method:** `experimental/runnables`
+
+**Request:**
+
+```typescript
+interface RunnablesParams {
+    textDocument: TextDocumentIdentifier;
+    /// If null, compute runnables for the whole file.
+    position?: Position;
+}
+```
+
+**Response:** `Runnable[]`
+
+```typescript
+interface Runnable {
+    label: string;
+    /// If this Runnable is associated with a specific function/module, etc, the location of this item
+    location?: LocationLink;
+    /// Running things is necessary technology specific, `kind` needs to be advertised via server capabilities,
+    // the type of `args` is specific to `kind`. The actual running is handled by the client.
+    kind: string;
+    args: any;
+}
+```
+
+rust-analyzer supports only one `kind`, `"cargo"`. The `args` for `"cargo"` look like this:
```

Comment:
Maybe it makes sense to distinguish different cargo commands?
I believe test targets, binaries and utility commands like 'check' should be different things. Anyway on the client side we have to filter them here and there.
```ts
export interface BinaryArgs { /*...*/ }
export interface TestArgs { /*...*/ }
export interface CheckArgs { /*...*/ }
export interface ExpandArgs { /*...*/ }

export interface Runnable {
    label: string;
    location?: lc.LocationLink;
    kind: "cargo";
    workspaceRoot?: string;
    args: BinaryArgs | TestArgs | CheckArgs | ExpandArgs;
}
```

Reply:
I’ve thought about this, but decided that looking at the cargoArgs[0] is probably good enough

Reply:
To me the main benefit of having different XXXArg kinds is that it would be easier and less error prone to get, for example, test id or expandable function name, etc. No need to parse raw string array and write client side tests :)

Reply:
Better use a discriminated union with `kind` discriminator. Types are our friends, not enemies...
`any` is generally bad practice and lsp has it just because it is legacy (e.g. `unknown` is called to replace it), bear with me...

Reply:
It can't be a discritimated union, because it needs to be open. We'll use `unknown` once the LSP spec uses unknown. 

I've though about it, and I see how for Cargo case we *might* want to use `BinaryArgs | TestArgs` in principle, but:

* it seems like we don't need those *right now*
* it seems like when we add support for non-cargo based build systems, the specific layout here will change.

So, I think I'd stick with the simplest possible thing for now.